### PR TITLE
Refactor `launch` for greater extensibility

### PR DESCRIPTION
**Pull request by Yard1**
This PR refactors `commands.launch` to break up the logic into smaller functions, allowing for the environment preparation logic and actual launch logic to be decoupled. Therefore, a developer can now prepare an environment dict (and command list, where applicable) without having to launch any processes.

There have been no changes to the logic - this is purely a refactor of an existing structure.

As discussed offline, this will unblock Ray Train and other distributed frameworks which launch their own jobs.

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks a lot for your PR! This looks great to me. Mind having a second look @muellerzr ?

Can you just solve the merge conflict @Yard1 ?

**Review by muellerzr — APPROVED** [review]
LG2M, great work! Once the merge conflict is fixed we can go ahead and bring it in 🤗

**Comment by Yard1** [comment]
@muellerzr Thanks, should be good now!

File: src/accelerate/utils/launch.py
Comment (dbpprt) [inline src/accelerate/utils/launch.py]: Ideally it should be imported from ..utils and not using the direct path to `.other`, following the same mechanism used for all other utils, isn't it?

Comment (Yard1) [reply src/accelerate/utils/launch.py]: That results in a circular import, unfortunately

Comment (dbpprt) [reply src/accelerate/utils/launch.py]: Got it, then LGTM.

**Comment by LouisCastricato** [comment]
FYI @natolambert this was the PR that i had mentioned to you ages ago. It'll be useful if you guys plan to do sweeps in TRL.

**Pull request merged**
